### ISSUE TITLE
Robotics tech is a silly intermediate requirement

### ISSUE THREAD

**Issue by dgw** [Warehousing/Warehousing#112]
In 2.0 (I haven't validated 1.1 or earlier), Robotics grants the robot frames only. To unlock passive provider and storage chests, either Construction Robotics or Logistic Robotics must then be researched.

This makes Robotics a silly prerequisite for Warehousing's first-tier logistic research, since it's still impossible to craft what it unlocks without those extra recipes.

Unless there's a hidden way to have some prerequisites treated as "OR" logic instead of "AND", the viable options are:

- Require both Construction Robotics and Logistic Robotics
- Require only Construction Robotics
- Require only Logistic Robotics

@Anoyomouse Want to have another chat with Nova about this one, see what they think?

**Comment by Anoyomouse** [comment]
So, chatted to Nova, and yes, this makes sense, we can follow what you need for logistics chests

One thought would be to unlock the "advanced" logistics warehouses (active provider, buffer, requester) by crafting a requester chest since you can have research unlock with items

But we're happy with your decision here :)

**Comment by dgw** [comment]
> One thought would be to unlock the "advanced" logistics warehouses (active provider, buffer, requester) by crafting a requester chest since you can have research unlock with items

This unfortunately doesn't do anything for this issue directly. `logistic-system` is a leaf research (no other techs declare it as a prerequisite [in `base`](https://github.com/wube/factorio-data/blob/bc5622fd4ccd60efe1f3b684d57d990604e1460e/base/prototypes/technology.lua#L3985-L3986)) and it is the only place where active/buffer/requester chest recipes are unlocked.

Could be worth thinking about, but for now I am focused on the "Robotics" problem. :)

> we can follow what you need for logistics chests

The problem is that the base game provides multiple techs that unlock those recipes, but doesn't also let technologies declare multiple paths to unlocking themselves (i.e. `warehousing-logistics-research-1` (WLR1 for short) can't say "my prerequisites are […], and EITHER construction-robotics OR logistic-robotics").

It's not actually possible to follow _exactly_ what the base game requires for logistic chests, because there are multiple paths. Same problem as iron-stick; we couldn't say "any of the techs that unlocks iron-stick should also unlock this research".

Because you suggested technology triggers, I had the idea to unlock WLR1 upon crafting either a Logistic or Construction robot, or upon building either a passive-provider-chest or storage-chest. Sadly, you can't do that; the game refuses to load the mod if `research_trigger` is not a plain dict defining just one trigger.

So I'm back to trying to think of ideas that aren't one of the three I outlined in the OP, because I don't really _like_ any of those. If I can't come up with anything better in a few days, I'll implement the first option (require both robot techs) and move on. 🙃

**Comment by Anoyomouse** [comment]
So, digging though the game files, you get:
Robotics -> Construction robotics (passive provider / storage chest)
Robotics -> Logistic robotics (passive provider / storage chest)
Logistic Robotics -> Logistic System (active provider / requester / buffer)

So, in my mind 2 approaches, well 3 ... 
- We could have it almost like it is with something that depends on logistic robotics,, which makes the most sense to me, and have it unlock all the warehouses
- We could have a tech that unlocks "nothing", but is a prerequisite for unlocking the warehouses once you craft the chests like:

Level 1:
```lua
name = "Logistic-warehousing"
prerequisites = {"robotics", "basic warehousing"}
 unlock = { type = "nothing" }
 requires = { ... } -- something along the lines of logistic robotics
```

Level 2:
```lua
prerequisites = {"logistic warehousing"}
research_trigger =  {
      type = "craft-item",
      item = "storage-chest"
    }
    unlock = { "passive provider warehouse", "storage warehouse" }
```

Level 3:
```lua
prerequisites = {"Logistic warehousing"}
research_trigger =  {
      type = "craft-item",
      item = "requester-chest"
    }
    unlock = { "active provider", "requester", "buffer" }
```

- Plan C is to inject warehousing into logistic robots and construction robots during the data-final-fixes .. but that's nasty, and feels worse than making something dependent on something

**Comment by dgw** [comment]
> * Plan C is to inject warehousing into logistic robots and construction robots during the data-final-fixes .. but that's nasty, and feels worse than making something dependent on something

Agree, I do not like the idea of data-fixes'ing or data-final-fixes'ing `base` stuff at all.

> * We could have a tech that unlocks "nothing", but is a prerequisite for unlocking the warehouses once you craft the chests like:

_(Not quoting the whole section for obvious reasons 🙃)_

This annoys me for the reason already stated: You _cannot_ have alternate paths to the same research, and which type of logistic chest would unlock a tier of logistic ware/storehouses simply has to be arbitrary. Why storage + requester? Why not passive + buffer? It doesn't lend itself well to a logical progression, IMO.

Another thing to consider for this is we'd need new technology icons, and also migrations. Not impossible (the migrations especially are easy) but a bigger change than I'd really like to make just to fix this relatively small quirk.

> * We could have it almost like it is with something that depends on logistic robotics,, which makes the most sense to me, and have it unlock all the warehouses

It would have to. In `base` the only tech with `logistic-robotics` as a prereq is `logistic-system`, which unlocks the rest of the chest types. While simple, we would lose the two-tier Warehousing logistic researches paralleling `base` tech and ultimately prevent players from accessing ware/storehouse versions of their chests for an artificially long time.

----

Honestly, I am still leaning toward requiring either _both_ Construction + Logistic Robotics techs, or _only_ Construction Robotics (because it is often what players go for first; having robots build stuff for you is generally practical sooner than having logi-bots run stuff around).

I guess what frustrates me the most is there being _no way_ to implement the Good Solution™ to this because of game (or mod API) limitations—short of janky `control.lua` logic stuff that wouldn't show up in the tech tree anyway.

In the process of getting early tester access to 2.0, did you make friends with anyone at Wube? Maybe ask if they'd now consider an "OR" mode for tech prereqs? 😁 I'm tempted to go post about it on the forums, but seeing that this idea came up already and got pretty much ignored each time (that I could find) is discouraging.

**Comment by dgw** [comment]
Having not made any progress on other alternatives that I like, I present #113 (simple option) and #114 (complicated, maybe-overengineered option).

Downside to #114 is it adds 8 (!) new strings that need to be localized. I groused in the PR about a missing game feature that would make translators' jobs a little easier, but doesn't exist.

**Comment by dgw** [comment]
> Why not make the tech unlock be this

Because I don't like `craft-item` unlocks, and they still have the same problem as prerequisites re: multiple pathways to the same unlock (read: you can't do it).

**Comment by SGZ-Creations** [comment]
ok
